### ISSUE TITLE
The output of Softmax layer is incorrect

### ISSUE THREAD

**Issue by WenguoLi** [alibaba/MNN#235]
I use softmax to normalize the output probability. I find that the output of MNN is not consistent with the output of tf.nn.softmax of tensorflow.  I think there are some problems with the implementation of softmax in MNN. Please check. Thanks!

**Comment by jxt1234** [comment]
What's the different rate? Can you show the value of mnn and tf.nn.softmax ?

**Comment by WenguoLi** [comment]
Thanks for your reply!
Results of mnn :
![mask_1](https://user-images.githubusercontent.com/31765154/63245475-e127b380-c292-11e9-9e1b-5fc898da2a76.png)
Result of tf.nn.softmax:
![mask_](https://user-images.githubusercontent.com/31765154/63245490-e84ec180-c292-11e9-99ec-d7f7489d2f42.png)
BRs

**Comment by yutianhang** [comment]
Can you provide the model causing problem and corresponding test code ?

**Comment by WenguoLi** [comment]
Sorry, I can't provide you with the model file for copyright reasons, but you can use the model of deeplabv3plus, just add Softmax after ResizeBilinear.

**Comment by WenguoLi** [comment]
I can get the right result by referring to the softMax implementation of [Paddle-Lite](https://github.com/Paddle Paddle/Paddle-Lite). The code is as follows:

```
#define c_exp_hi 88.3762626647949f
#define c_exp_lo -88.3762626647949f

#define c_cephes_LOG2EF 1.44269504088896341
#define c_cephes_exp_C1 0.693359375
#define c_cephes_exp_C2 -2.12194440e-4

#define c_cephes_exp_p0 1.9875691500E-4
#define c_cephes_exp_p1 1.3981999507E-3
#define c_cephes_exp_p2 8.3334519073E-3
#define c_cephes_exp_p3 4.1665795894E-2
#define c_cephes_exp_p4 1.6666665459E-1
#define c_cephes_exp_p5 5.0000001201E-1

/* exp() computed for 4 float at once */
static inline float32x4_t exp_ps(float32x4_t x) {
    float32x4_t tmp, fx;

    float32x4_t one = vdupq_n_f32(1);
    x = vminq_f32(x, vdupq_n_f32(c_exp_hi));
    x = vmaxq_f32(x, vdupq_n_f32(c_exp_lo));

    /* express exp(x) as exp(g + n*log(2)) */
    fx = vmlaq_f32(vdupq_n_f32(0.5f), x, vdupq_n_f32(c_cephes_LOG2EF));

    /* perform a floorf */
    tmp = vcvtq_f32_s32(vcvtq_s32_f32(fx));

    /* if greater, substract 1 */
    uint32x4_t mask = vcgtq_f32(tmp, fx);
    mask = vandq_u32(mask, vreinterpretq_u32_f32(one));

    fx = vsubq_f32(tmp, vreinterpretq_f32_u32(mask));

    tmp = vmulq_f32(fx, vdupq_n_f32(c_cephes_exp_C1));
    float32x4_t z = vmulq_f32(fx, vdupq_n_f32(c_cephes_exp_C2));
    x = vsubq_f32(x, tmp);
    x = vsubq_f32(x, z);

    static const float cephes_exp_p[6] = {c_cephes_exp_p0, c_cephes_exp_p1,
                                        c_cephes_exp_p2, c_cephes_exp_p3,
                                        c_cephes_exp_p4, c_cephes_exp_p5};
    float32x4_t y = vld1q_dup_f32(cephes_exp_p + 0);
    float32x4_t c1 = vld1q_dup_f32(cephes_exp_p + 1);
    float32x4_t c2 = vld1q_dup_f32(cephes_exp_p + 2);
    float32x4_t c3 = vld1q_dup_f32(cephes_exp_p + 3);
    float32x4_t c4 = vld1q_dup_f32(cephes_exp_p + 4);
    float32x4_t c5 = vld1q_dup_f32(cephes_exp_p + 5);

    y = vmulq_f32(y, x);
    z = vmulq_f32(x, x);

    y = vaddq_f32(y, c1);
    y = vmulq_f32(y, x);
    y = vaddq_f32(y, c2);
    y = vmulq_f32(y, x);
    y = vaddq_f32(y, c3);
    y = vmulq_f32(y, x);
    y = vaddq_f32(y, c4);
    y = vmulq_f32(y, x);
    y = vaddq_f32(y, c5);

    y = vmulq_f32(y, z);
    y = vaddq_f32(y, x);
    y = vaddq_f32(y, one);

    /* build 2^n */
    int32x4_t mm;
    mm = vcvtq_s32_f32(fx);
    mm = vaddq_s32(mm, vdupq_n_s32(0x7f));
    mm = vshlq_n_s32(mm, 23);
    float32x4_t pow2n = vreinterpretq_f32_s32(mm);

    y = vmulq_f32(y, pow2n);
    return y;
}

#if defined(__ARM_NEON) || defined(__ARM_NEON__)
#ifndef __aarch64__
static inline float32_t vmaxvq_f32(const float32x4_t &r) {
    float32x2_t v = vmax_f32(vget_high_f32(r), vget_low_f32(r));
    return vget_lane_f32(vpmax_f32(v, v), 0);
}

static inline float32_t vaddvq_f32(const float32x4_t &r) {
    float32x2_t v = vadd_f32(vget_high_f32(r), vget_low_f32(r));
    return vget_lane_f32(vpadd_f32(v, v), 0);
}
#endif  // __aarch64__
#endif  // __ARM_NEON__


static float find_max(const float *input, const int num_classes) {
    int remain = num_classes;
    float max = -std::numeric_limits<float>::max();

#if defined(__ARM_NEON) || defined(__ARM_NEON__)
    int loop = num_classes >> 3;
    remain = num_classes & 0x7;
    float32x4_t __max = vdupq_n_f32(max);
    for (int i = 0; i < loop; ++i, input += 8) {
        float32x4_t x0 = vld1q_f32(input);
        float32x4_t x1 = vld1q_f32(input + 4);
        __max = vmaxq_f32(x0, __max);
        __max = vmaxq_f32(x1, __max);
    }
    max = vmaxvq_f32(__max);
#endif

    for (int i = 0; i < remain; ++i) {
        max = std::max(max, input[i]);
    }
    return max;
}

static void SoftmaxBasic(const float *input, int num_classes, float *y) {
    float *output = y;
    // find max
    float max = find_max(input, num_classes);

    // exp(x - max) and sum(exp(x - max))
    int remain = num_classes;
    float sum = 0.f;
    
#if defined(__ARM_NEON) || defined(__ARM_NEON__)
    int loop = num_classes >> 3;
    remain = num_classes & 0x7;
    float32x4_t __max = vdupq_n_f32(max);
    float32x4_t __sum = vdupq_n_f32(0.f);
    for (int i = 0; i < loop; ++i, input += 8, output += 8) {
        float32x4_t x0 = vld1q_f32(input);
        float32x4_t x1 = vld1q_f32(input + 4);
        x0 = vsubq_f32(x0, __max);
        x1 = vsubq_f32(x1, __max);
        x0 = exp_ps(x0);
        x1 = exp_ps(x1);
        __sum = vaddq_f32(x0, __sum);
        __sum = vaddq_f32(x1, __sum);
        vst1q_f32(output, x0);
        vst1q_f32(output + 4, x1);
    }
  sum += vaddvq_f32(__sum);
#endif  // __ARM_NEON__

    for (int i = 0; i < remain; ++i) {
        float out = expf(input[i] - max);
        sum += out;
        output[i] = out;
    }

    // exp(x - max) / sum
    float inv_sum = 1.f / sum;
    output = y;
#if defined(__ARM_NEON) || defined(__ARM_NEON__)
    float32x4_t __inv_sum = vdupq_n_f32(inv_sum);
    for (int i = 0; i < loop; ++i, output += 8) {
        float32x4_t x0 = vld1q_f32(output);
        float32x4_t x1 = vld1q_f32(output + 4);
        x0 = vmulq_f32(x0, __inv_sum);
        x1 = vmulq_f32(x1, __inv_sum);
        vst1q_f32(output, x0);
        vst1q_f32(output + 4, x1);
    }
#endif
    for (int i = 0; i < remain; ++i) {
        output[i] *= inv_sum;
    }
}

static void Softmax(const float* src, float* dst, int batch, int length, int num_classes)
{
    for(int k = 0; k<batch; k++)
        for(int i=0; i<length; i++)
        {
            int offset = (k*length + i)*num_classes;
            SoftmaxBasic(src + offset, num_classes, dst + offset);
        }
}
```

**Comment by jxt1234** [comment]
What's your code for translate softmax's value to picture?

**Comment by WenguoLi** [comment]
(a > = b ? 1: 0)  * 255.0

**Comment by jxt1234** [comment]
I can't reproduce it by add softmax layer after resizebilinear. Can you assure your code is updated?

Or upload the input tensor of softmax and text and after

**Comment by WenguoLi** [comment]
Resizebilinear tensor data has been posted to the attachment.
Shape: [ 1, 512, 512, 2]
[ResizeBilinear.zip](https://github.com/alibaba/MNN/files/3518789/ResizeBilinear.zip)

**Comment by jxt1234** [comment]
e...... Could you should your full code of post treat?

If you use the output tensor directly, you should write as this because Resizebilinear-> Softmax tensor will use NC4HW4:
mask[i] = (uint8_t)(pbsPtr[4i+1] > pbsPtr[4i] ? 255 : 0);

**Comment by WenguoLi** [comment]
Post-processing is very simple. 

```
static int docodePortrait(const Tensor* probs, uint8_t *mask)
{ 
    const int channel  = probs->channel();
    const int height   = probs->height();
    const int width    = probs->width();
    auto pbsPtr        = probs->host<float>();
    //auto pbsPtr        = probs->host<int>();
    
    std::cout << "channel = " << channel << std::endl;
    std::cout << "height = " << height << std::endl;
    std::cout << "width = " << width << std::endl;
    
    const int number_of_pixels = height*width;
    for (int i = 0; i < number_of_pixels; i++) { 
        //mask[i] = (uint8_t)(pbsPtr[number_of_pixels+i] > pbsPtr[i] ? 255 : 0);
        mask[i] = (uint8_t)(pbsPtr[2*i+1] > pbsPtr[2*i] ? 255 : 0);
        //mask[i] = (uint8_t)(pbsPtr[i]* 255);
    }   
    return 0;
}
```

```
    // get output
    auto probs        = mnnNet->getSessionOutput(session, "ResizeBilinear");//"Softmax");//"ArgMax");//
    
    std::vector<int> out_shape = probs->shape();
    std::cout << "output: ( " << out_shape[0] << ", " << out_shape[1] << "," << out_shape[2] << "," << out_shape[3] << " )"<< std::endl;    
    
    Tensor probsHost(probs, Tensor::TENSORFLOW);
    probs->copyToHostTensor(&probsHost);

    //read one bmp
    std::ofstream file("./ResizeBilinear.bin", std::ios::out | std::ios::binary);
    if (!file) {
        return -1;
    }
    
    file.write(reinterpret_cast<const char*>(probsHost.host<float>()), probsHost.size());
    
    file.close();

    std::vector<uint8_t> masks(targetWidth*targetHeight);    
    docodePortrait(&probsHost, masks.data());

    stbi_write_png(outputImageFileName, targetWidth, targetHeight, 1, masks.data(), targetWidth);
    stbi_image_free(inputImage);
```

**Comment by jxt1234** [comment]
What's your version of mnn ? I use the data and test in pc and mobile. The picture is all ok.
![mask](https://user-images.githubusercontent.com/5484403/63323519-940e1500-c358-11e9-8cc0-ec5a035e26b8.png)

**Comment by WenguoLi** [comment]
The file I gave you is output of ResizeBilinear. You need to do Softmax and then output it. Is that so?
version of mnn : 0.2.0.6

**Comment by WenguoLi** [comment]
I pull the newest version of MNN,  Softmax output is still wrong.
```
    auto probs        = mnnNet->getSessionOutput(session, "Softmax");
    
    std::vector<int> out_shape = probs->shape();
    std::cout << "output: ( " << out_shape[0] << ", " << out_shape[1] << "," << out_shape[2] << "," << out_shape[3] << " )"<< std::endl;    
    
    Tensor probsHost(probs, Tensor::TENSORFLOW);
    probs->copyToHostTensor(&probsHost);

    std::vector<uint8_t> masks(targetWidth*targetHeight);    
    docodePortrait(&probsHost, masks.data());

    stbi_write_png(outputImageFileName, targetWidth, targetHeight, 1, masks.data(), targetWidth);
    stbi_image_free(inputImage);

```
![mask](https://user-images.githubusercontent.com/31765154/63324976-c8370500-c35b-11e9-9408-5eb7dca2483a.png)

**Comment by jxt1234** [comment]
> The file I gave you is output of ResizeBilinear. You need to do Softmax and then output it. Is that so?

I use the output and do softmax on it. And the result is right.

**Comment by WenguoLi** [comment]
The model runs with 4 threads. Can you verify the multi-threaded softmax?

**Comment by WenguoLi** [comment]
backend :  pc cpu  and  android arm

**Comment by WenguoLi** [comment]
Can you share the code of doing Softmax  ?

**Comment by jxt1234** [comment]
Write like test/op/SoftmaxTest.cpp.

**Comment by WenguoLi** [comment]
please verify input with CAFFE format.
[ResizeBilinear_caffe.zip](https://github.com/alibaba/MNN/files/3519471/ResizeBilinear_caffe.zip)

**Comment by jxt1234** [comment]
Could you give me a model to reproduce it . For the model come from Tensorflow, MNN's tensor in session will not use NCHW format, so the softmax  will not be done in CAFFE format.
